### ISSUE TITLE
Classes

### ISSUE THREAD

**Issue by arm61** [pythoninchemistry/intro_python_chemists#24]
Write a section introducing classes:

- class variables
- class methods/functions
- the `@property` decorator
- docstrings for classes

**Comment by RMeli** [comment]
@arm61 I can take this on with #23 if you want. I'll come up with a first ToC in the coming days so that we can agree on the general outline.

**Comment by RMeli** [comment]
Attributes, methods and docstrings for classes are discussed in #65. 

I think the `@property` decorator can be easily introduced to extend the `Molecule` class further (for example by checking that all the element numbers are valid in the setter), in an advanced section.

**Comment by RMeli** [comment]
I suggest the following discussion for the advanced section:

* `@property` decorator
  * Use to set `self.elements` attribute with valid elements only
* Inheritance
  * Linked to #65: `Molecule` inherit from `SimpleMolecule` to avoid duplication of code (`__init__` method)

Please let me know if other useful concepts should be added.

**Comment by RMeli** [comment]
Scratch that. What about a short introduction to `networkx` and show inheritance by deriving a molecule from `networkx.Graph`?

**Comment by arm61** [comment]
So using a graph network to describe a molecule?

**Comment by RMeli** [comment]
Yes, something like [A Minimal Molecule API](https://depth-first.com/articles/2020/04/06/a-minimal-molecule-api/), but simplified.

**Comment by arm61** [comment]
Not opposed to the idea. In particular, cause so much of Python is about building on top of other packages.

**Comment by RMeli** [comment]
Yes, that's what I was thinking. Instead of creating something from scratch to describe inheritance I think it's good to show that you can actually inherit from third-party libraries and avoid re-inventing the wheel. 

This week I'm very busy, but I'll sketch out the advanced OOP section the coming weekend or next week.

**Comment by RMeli** [comment]
I was thinking of the following for the advanced section:

- [ ] Extension for the `Molecule` class
  - [ ] Introduce more dunder methods
    - [ ] `__repr__`
    - [ ] `__len__` and `__getitem__` to get atoms
    - [ ] `__iter__` to loop over atoms (maybe too advanced if needs to introduce iterators?)
  - [ ] Introduce decorators and `@property`
- [ ] Introduction to inheritance
  - [ ] Inherit a molecule from `networkx.Graph`
  - [ ] Explain the [drawbacks](https://medium.com/@cscalfani/goodbye-object-oriented-programming-a59cda4c0e53) of this approach

Do you have any other suggestion on what should (or shouldn't) be included?
